pandas/core/index.py

```diff
@@ -1753,6 +1775,14 @@ def _get_nearest_indexer(self, target, limit):
         indexer = np.where(op(left_distances, right_distances)
                            | (right_indexer == -1),
                            left_indexer, right_indexer)
+        if tolerance is not None:
+            indexer = self._limit_indexer_tolerance(
+                target, indexer, tolerance)
+        return indexer
+
+    def _limit_indexer_tolerance(self, target, indexer, tolerance):
+        distance = abs(self.values[indexer] - target)
+        indexer = np.where(distance <= tolerance, indexer, -1)
```

Comment:
if the distance and the tolerance are invalid units, e.g. an integer on a ``DatetimeIndex`` this will give a very odd error message. Further this is not tested AFAICT.